guides/data_modelling/cross_context_boundaries.md

```diff
@@ -0,0 +1,563 @@
+
+With our routes in place, let's add the ability to add an item to our cart from the product show page. Create a new file at `lib/hello_web/controllers/cart_item_controller.ex` and key this in:
+
+```elixir
+defmodule HelloWeb.CartItemController do
+  use HelloWeb, :controller
+
+  alias Hello.ShoppingCart
+
+  def create(conn, %{"product_id" => product_id}) do
+    case ShoppingCart.add_item_to_cart(conn.assigns.current_scope, conn.assigns.cart, product_id) do
+      {:ok, _item} ->
+        conn
+        |> put_flash(:info, "Item added to your cart")
+        |> redirect(to: ~p"/cart")
+
+      {:error, _changeset} ->
+        conn
+        |> put_flash(:error, "There was an error adding the item to your cart")
+        |> redirect(to: ~p"/cart")
+    end
+  end
+
+  def delete(conn, %{"id" => product_id}) do
+    {:ok, _cart} = ShoppingCart.remove_item_from_cart(conn.assigns.current_scope, conn.assigns.cart, product_id)
+    redirect(conn, to: ~p"/cart")
+  end
+end
+```
+
+We defined a new `CartItemController` with the create and delete actions that we declared in our router. For `create`, we call a `ShoppingCart.add_item_to_cart/2` function which we'll implement in a moment. If successful, we show a flash successful message and redirect to the cart show page; else, we show a flash error message and redirect to the cart show page. For `delete`, we'll call a `remove_item_from_cart` function which we'll implement on our `ShoppingCart` context  and then redirect back to the cart show page. We haven't implemented these two shopping cart functions yet, but notice how their names scream their intent: `add_item_to_cart` and `remove_item_from_cart` make it obvious what we are accomplishing here. It also allows us to spec out our web layer and context APIs without thinking about all the implementation details at once.
+
+Let's implement the new interface for the `ShoppingCart` context API in `lib/hello/shopping_cart.ex`:
+
+```diff
++  alias Hello.Catalog
+-  alias Hello.ShoppingCart.Cart
++  alias Hello.ShoppingCart.{Cart, CartItem}
+   alias Hello.Accounts.Scope
+
++  def get_cart(%Scope{} = scope) do
++    Repo.one(
++      from(c in Cart,
++        where: c.user_id == ^scope.user.id,
++        left_join: i in assoc(c, :items),
++        left_join: p in assoc(i, :product),
++        order_by: [asc: i.inserted_at],
++        preload: [items: {i, product: p}]
++      )
++    )
++  end
+
+   def create_cart(%Scope{} = scope, attrs \\ %{}) do
+     with {:ok, cart = %Cart{}} <-
+            %Cart{}
+            |> Cart.changeset(attrs, scope)
+            |> Repo.insert() do
+       broadcast(scope, {:created, cart})
+-      {:ok, cart}
++      {:ok, get_cart!(scope, cart.id)}
```

Comment:
```suggestion
+      {:ok, get_cart(scope, cart.id)}
```

I think I did that wrong in the scopes PR, looks like the function doesn't use `!`.